google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceInfo.java

```diff
@@ -35,13 +36,16 @@ public enum InstanceField implements FieldSelector {
     DISPLAY_NAME("display_name"),
     NODE_COUNT("node_count"),
     PROCESSING_UNITS("processing_units"),
+    AUTOSCALING_CONFIG("autoscaling_config"),
     LABELS("labels");
 
     static InstanceField[] defaultFieldsToUpdate(InstanceInfo info) {
-      if (info.getNodeCount() > 0) {
-        return new InstanceField[] {DISPLAY_NAME, NODE_COUNT, LABELS};
+      if (info.getAutoscalingConfig() != null) {
+        return new InstanceField[] {DISPLAY_NAME, AUTOSCALING_CONFIG, LABELS};
+      } else if (info.getNodeCount() > 0) {
+        return new InstanceField[] {DISPLAY_NAME, AUTOSCALING_CONFIG, NODE_COUNT, LABELS};
```

Comment:
Why does this array contain  both `NODE_COUNT` and `AUTOSCALING_CONFIG` ? It should contain just `NODE_COUNT` ? Could you check if any unit tests broke here or this was a silent error?

Reply:
There are a few edge cases and open question here:

1. when `getAutoscalingConfig != null`, we are updating just `AUTOSCALING_CONFIG`. Shouldn't we have checks to include `NODE_COUNT` and `PROCESSING_UNITS` (to be set to 0) ?
2. If `getNodeCount > 0`, we are updating just `NODE_COUNT`. What if an instance already had `PROCESSING_UNITS` ? I assume the backend will throw a validation error. Similar question for the `else` condition. 
3. If we are not marking `PROCESSING_UNITS` or `NODE_COUNT` as 0 in the second check, then why are we explicitly marking `AUTOSCALING_CONFIG` as null?

Reply:
For #1, customers can only autoscaling_config once autoscaler is enabled. node_count and processing_units are output_only params. An instance can still have node_count and processing_units, it is just customers cannot modify them. This is consistent with CBT autoscaling API: https://screenshot.googleplex.com/6mi2r2QUKMfTk7x

For #2, this seems to be the logic before the autoscaler related changes. My understanding is that this method is to provide a default field mask when user did not specify one. I'm not sure why we did this in the first place -- as you pointed out, if an instance is specified with both nodes and processing units, this can be wrong. I'm following an existing pattern here.

For #3, if the autoscaling_config is not null, that means this is an autoscaler instance, and users should not specify processing_units or node_count as part of the field mask. If it is null, then the user intention will be to update a non-autoscaling instance.